pkg/util/nativeutils/republisher/republisher.go

```diff
@@ -0,0 +1,63 @@
+package republisher
+
+import (
+	"bytes"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+)
+
+type Validator func(bytes.Buffer) error
+
+// Republisher handles the repropagation of Agreement messages
+type Republisher struct {
+	tpc        topics.Topic
+	broker     eventbus.Broker
+	id         uint32
+	validators []Validator
+}
+
+// New creates a Republisher
+func New(eb eventbus.Broker, tpc topics.Topic, v ...Validator) *Republisher {
+	r := &Republisher{
+		broker:     eb,
+		tpc:        tpc,
+		validators: v,
+	}
+	r.id = r.Activate()
+	return r
+}
+
+// Stop a Republisher
+func (r *Republisher) Stop() {
+	r.broker.Unsubscribe(r.tpc, r.id)
+}
+
+// Activate the Republisher by listening to Agreement messages through a
+// callback listener
+func (r *Republisher) Activate() uint32 {
+	if r.id != 0 {
+		return r.id
+	}
+
+	l := eventbus.NewCallbackListener(r.Republish)
+	r.id = r.broker.Subscribe(r.tpc, l)
+	return r.id
+}
+
+// Republish intercepts an Agreement message and repropagates it immediately
+// after applying any eventual validation logic
+func (r *Republisher) Republish(b bytes.Buffer) error {
```

Comment:
nitpick: Is the republisher intended only for use upon Agreement messages? It's also used for topics.Candidate and topics.Reduction.

Reply:
As far as I could tell, these are included. Check the reduction factory and candidate broker, and they should be there.

Reply:
no, indeed it is a copy/paste error